aten/src/ATen/native/Normalization.cpp

```diff
@@ -615,4 +635,21 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_backward_cpu(const Tensor& grad_ou
     });
 }
 
+TORCH_IMPL_FUNC(renorm_out)(const Tensor& self, const Scalar& p, int64_t dim,
+                            const Scalar& maxnorm, const Tensor& out) {
+  auto self_sizes = self.sizes();
+  dim = c10::maybe_wrap_dim(dim, self_sizes.size());
+  DimVector row_sizes(self_sizes.size(), 1);
+  row_sizes[dim] = self_sizes[dim];
+
+  DimVector reduce_dims(self_sizes.size());
+  std::iota(reduce_dims.begin(), reduce_dims.end(), 0);
+  reduce_dims.erase(reduce_dims.begin() + dim);
+  auto norm = at::linalg_vector_norm(self, p.toDouble(), reduce_dims, /*keepdim=*/true);
```

Comment:
for fp16, norm has a good chance of overflowing. In previous implementation the result was kept in AccT, perhaps makes sense to do it here. Annoyingly, I think our fused implementation of half norm producing float outputs is broken, so it would cast inputs to float first, but it's safer

Reply:
Done, cuda-half is now around 5 us slower for small sizes but the larger size it's still actually faster than the THC kernel.